SC-2173 Introduce new component for SC new navigation

gregaubert opened the pull request:
Some upgraded popup to be able to use them anywhere like tooltips. Maybe I could have done it directly in popup but I was a bit too afraid of breaking too many stuff...

**Checklist**

* [ ] Functional validation
* [ ] Documentation update
* [ ] Update changelog

gregaubert committed:
SC-2173 Introduce PortalPopup component

gregaubert commented:
@wouter-admiraal-sonarsource Pinging you because I think I saw that you worked quite a lot on the tooltips stuff

antonio-garcia-sonarsource commented on components/ui/PortalPopup.tsx:
```diff
@@ -0,0 +1,228 @@
+  }
+}
+
+class PopupPortal extends React.Component {
```
Maybe we can find a better name for this? I found confusing having PortalPopup and PopupPortal classes here. 

Otherwise, It looks good to me!

gregaubert replied:
Ah yeah indeed, I didn't notice. Maybe I could rename it to something like: "PortalWrapper" or "PortalPopupWrapper", would it be better ? Or do you have something in mind?

antonio-garcia-sonarsource replied:
Yes, I would say `PortalWrapper` fits perfectly here

gregaubert committed:
SC-2173 Add new cog icon and new animated arrow icon

gregaubert commented:
@antonio-garcia-sonarsource I also moved some new icons from SC to here.

@wouter-admiraal-sonarsource Do you think I should also move this component here: https://github.com/SonarSource/sonarcloud-core/pull/1132/commits/5977b4bf5b977d35c8bf98aa238a19843a7c1432 ?

siegfried-ehret-sonarsource commented on components/ui/PortalPopup.tsx:
```diff
@@ -0,0 +1,228 @@
+  return state.height !== undefined;
+}
+
+export default class PortalPopup extends React.Component<PopupProps, State> {
```
Not sure if that is useful but you can remove `children` from the PopupProps and use `react.PropsWithChildren<Props>` here.

gregaubert replied:
Yes, I could but I don't want the children to be optional like in `React.PropsWithChildren`, so I just thought it was more straightforward to define it myself...

wouter-admiraal-sonarsource commented on components/controls/Tooltip.css:
```diff
@@ -23,9 +23,6 @@
   box-sizing: border-box;
-}
-
-.tooltip {
```
Bold move 😎 .

wouter-admiraal-sonarsource commented on components/icons/ArrowIcon.tsx:
```diff
@@ -0,0 +1,50 @@
+export default function ArrowIcon({
+  animated = false,
+  fill = 'currentColor',
+  revert = false,
```
NP: I'm not sure `revert` is intuitive here (I had to try it out to understand the intent). Maybe `leftToRight` would be more comprehensible?

gregaubert replied:
WDYT about `inverseDirection` ?

wouter-admiraal-sonarsource commented on components/ui/PortalPopup.tsx:
```diff
@@ -0,0 +1,228 @@
+    return (
+      <Popup
+        arrowStyle={
+          isMeasured(this.state)
```
NP: You're calling this `isMeasured()` twice. You could also just check `style`:

```tsx
        arrowStyle={
         style
            ? this.adjustArrowPosition(placement, { leftFix, topFix })
            : undefined
        }
```

wouter-admiraal-sonarsource commented on components/ui/PortalPopup.tsx:
```diff
@@ -0,0 +1,228 @@
+          left = toggleRect.left + toggleRect.width - width;
+          top = toggleRect.top + toggleRect.height;
+          break;
+        case PopupPlacement.LeftTop:
+          left = toggleRect.left - width;
+          top = toggleRect.top;
+          break;
```
Not sure all positionings are correct. See this for `LeftTop`:

<img width="480" alt="Screenshot 2020-07-09 at 10 46 35" src="https://user-images.githubusercontent.com/45544358/87018464-91671900-c1d1-11ea-82ac-d6459c2d3758.png">

Could be due to the surrounding elements, which might confuse `ScreenPositionFixer`, but it's worth to double check.

gregaubert replied:
I just checked again and I think it's correct. Here too, the top of the popup is on the left and is aligned with the top of your text. You'd get the same result if you used a normal popup.

wouter-admiraal-sonarsource commented on components/ui/PortalPopup.tsx:
```diff
@@ -0,0 +1,228 @@
+  return state.height !== undefined;
+}
+
+export default class PortalPopup extends React.Component<Props, State> {
```
NP: I wonder if this component shouldn't be added to the `popups.tsx` file, instead of living on its own. It's basically a more advanced variant of the `Popup` component, not its own thing. This feels a bit like the Buttons, which are all in the same file, regardless of their type.

wouter-admiraal-sonarsource commented on CHANGELOG.md:
```diff
@@ -2,6 +2,12 @@
 ## Unreleased
 
+- Add right-bottom popup position
+- Allow to pass ref to popup
```
NP: 

> ...to Popup

Camelcase? So we know we're talking about the component class.

wouter-admiraal-sonarsource commented on CHANGELOG.md:
```diff
@@ -2,6 +2,12 @@
+- Add right-bottom popup position
+- Allow to pass ref to popup
+- Add PortalPopup component: allow to create popup outside it's overflow:hidden parent
+- Add animated Arrow icon
+- Add cog icon
```
NP: either lower case "Arrow", or camelcase "cog" 🙂 .

wouter-admiraal-sonarsource reviewed (approved):
LGTM 👍 . Some small nitpicks. Also, maybe double check the positioning works as intended.

wouter-admiraal-sonarsource commented:
> @wouter-admiraal-sonarsource Do you think I should also move this component here: [SonarSource/sonarcloud-core@5977b4b](https://github.com/SonarSource/sonarcloud-core/commit/5977b4bf5b977d35c8bf98aa238a19843a7c1432) ?

That looks pretty cool. Yes please 🙂 . Would be nice if we can get those tests to work, though. Need help (rubber duck) with that?

gregaubert committed:
SC-2173 Add AutoEllipsis component

gregaubert committed:
SC-2173 Add helpers to easily get theme values in emotion components

gregaubert commented:
@wouter-admiraal-sonarsource , @antonio-garcia-sonarsource I applied all feedbacks and moved the autoellipsis component here and was able to cover it properly with tests. I also added some theme helpers to be used with emotion components.

antonio-garcia-sonarsource reviewed (approved):
LGTM!

gregaubert commented:
Ok I'll merge and proceed with the release so that I unlock my PR on SC side... @wouter-admiraal-sonarsource feel free to still add comments if you do a last review on the two last commit and find anything!

wouter-admiraal-sonarsource commented on components/theme.ts:
```diff
@@ -49,5 +49,17 @@ export const ThemeProvider = EmotionThemeProvider as React.ProviderExoticCompone
 >;
 export const useTheme = emotionUseTheme as () => Theme;
 
+export function themeGet(type: keyof Theme, name: keyof Theme['sizes']) {
```
NP: technically, this `name` param should probably have been typed as `string` rather than `keyof Theme['sizes']` 🤷‍♂️ .

wouter-admiraal-sonarsource replied:
And perhaps it doesn't need exporting?

gregaubert replied:
It's good to have it exported, I use it already in a few places to get zIndexes. It's kind of the generic helper do get theme data we don't use often...

wouter-admiraal-sonarsource commented on components/ui/__tests__/AutoEllipsis-test.tsx:
```diff
@@ -0,0 +1,76 @@
+  );
+
+  expect(wrapper).toMatchSnapshot();
+  expect(wrapper.find('span').hasClass('medium')).toBe(true);
```
NP: I guess this one is redundant, if you have the snapshot.

wouter-admiraal-sonarsource commented:
Just added some NPs. And because it's already merged, feel free to ignore them 😉 .
